rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2594 @@
+        {
+          "properties": {
+            "oneOf": {
+              "type": "array",
+              "minItems": 1,
+              "items": {
+                "oneOf": [
+                  { "$ref": "#/$defs/propertyTypeObject" },
+                  { "$ref": "#/$defs/dataTypeReference" },
+                  { "$ref": "#/$defs/propertyValues" }
+                ]
+              }
+            }
+          },
+          "required": ["oneOf"]
+        },
+        {
+          "properties": {
+            "type": { "const": "array" },
+            "items": { "$ref": "#/$defs/propertyValues" },
+            "minItems": { "type": "number" },
+            "maxItems": { "type": "number" }
+          },
+          "required": ["type", "items"]
+        }
+      ]
+    },
+    "propertyTypeObject": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "type": { "const": "object" },
+        "properties": {
+          "type": "object",
+          "propertyNames": {
+            "$comment": "Property names must be a valid URI to a Property Type",
+            "type": "string",
+            "format": "uri"
+          },
+          "patternProperties": {
+            ".*": {
+              "oneOf": [
+                { "$ref": "#/$defs/propertyTypeReference" },
+                {
+                  "type": "object",
+                  "properties": {
+                    "type": { "const": "array" },
+                    "items": { "$ref": "#/$defs/propertyTypeReference" },
+                    "minItems": { "type": "number" },
+                    "maxItems": { "type": "number" }
+                  },
+                  "required": ["type", "items"],
+                  "additionalProperties": false
+                }
+              ]
+            }
+          }
+        },
+        "required": {
+          "type": "array",
+          "items": "string"
+        }
+      }
+    },
+    "propertyTypeReference": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "$ref": {
+          "$comment": "Property Object values must be defined through references to the same valid URI to a Property Type",
+          "type": "string",
+          "format": "uri"
+        }
+      },
+      "required": ["$ref"]
+    },
+    "dataTypeReference": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "$ref": {
+          "type": "string",
+          "format": "uri"
+        }
+      },
+      "required": ["$ref"]
+    }
+  }
+}
+```
+
+### Defining the possible values of a property
+
+A Property Type describes its possible values through a `oneOf` definition, which can include:
+
+#### Data Types
+
+**Example 1**
+
+The `Favorite Quote` Property Type could define its value as being an instance of the `Text` Data Type.
+
+```json
+{
+  "kind": "propertyType",
+  "$id": "https://blockprotocol.org/types/@alice/property-type/favorite-quote",
```

Comment:
Sticking the comments here @nonparibus so we can reply in a chain because commenting directly on the PR is a mess (thanks GitHub 🙃 )

> 
> General nit 1: it's tautological to have both `types` and `[type]-type` in the URL slug here. Let's simplify `data-type` to `data`, along with the other types.
> 
> General nit 2: the namespace should come before `types` in the slug to match the public site.

> In this example, the new slug would be `https://blockprotocol.org/@blockprotocol/types/data/number`
> 
> In addition to `data`, orgs would also have:
> - `https://blockprotocol.org/@blockprotocol/types/entity`
> - `https://blockprotocol.org/@blockprotocol/types/link`
> - `https://blockprotocol.org/@blockprotocol/types/property`

Reply:
The URLs have always meant to be placeholders (although included to be indicative that URIs are verbose).

Thanks for highlighting. When it comes to the implementation we should decide on appropriate URI patterns

Reply:
Apologies @Alfred-Mountfield, @benwerner01 has previously flagged this weird GH behavior to me. My bad.

Reply:
You think that would be better than...
- `https://blockprotocol.org/@namespace/type/data/...`
- `https://blockprotocol.org/@namespace/type/property/...`
- `https://blockprotocol.org/@namespace/type/entity/...`
- `https://blockprotocol.org/@namespace/type/link/...`

Reply:
I'm sort of inclined to go for something else, where I don't have a concrete suggestion yet. My thought is that we probably don't want service-specific things just living in plain paths on blockprotocol.org. If this RFC gets merged with the Type System _as part of the `graph` service_, then I think we should probably look at having `graph` or `service/graph` or something in the path as well.

I do much prefer `data-type` over `types/data` too.

Reply:
The `/type/property/` pattern feels a little like browsing a tree (sort of a file system), which could imply that there will be more kinds after the `/types/` section.

The `/entity-type/` is more clear about the limited kinds we're dealing with.

I don't have a strong preference about what the naming should look like, just sharing my initial thoughts on the two variants.

Reply:
@nonparibus the benefit I see is whether we're talking about data (entities, properties, etc) or types (entity-types, property-types, etc) the URI has a consistent structure:
- `https://blockprotocol.org/[namespace]/[what is it]/[identifier]`
- `https://blockprotocol.org/@alice/entity-type/Person`
- `https://blockprotocol.org/@alice/entity/Elon-Musk`


But I don't have strong feelings about this so happy to go either way

Reply:
Yes but to be fair `[what is it]` doesn't **have** to be a single part of the path, it could be a few things like `types/entity` or `services/graph/entity-type` or something else

Reply:
I think having `service/graph` would be overkill, because we can assume that any merged thing represented in the spec is part of the core or a `service`.

I am arguing for `https://blockprotocol.org/[namespace]/[what is it]/[identifier]` wherein `[what is it]` is simply `type/:type:`, because we have types _of_ types (entity/property/link/data). I do not care too much if we think `entity-type` is better than `type/entity`, although because the latter better represents a _type of type_ it seems to provide more descriptive information per-bit (at least assuming my conception of types is correct).

Reply:
I think there is some value in considering namespacing according to the service so that we don't have to worry about overlapping things, (perhaps some event service comes up with a need for `type` or whatever). I don't feel particularly strongly about it.

I don't think _type of types_ is incorrect but perhaps its easiest for us to call them "classes of types".

My main thing about `entity-type` v `types/entity` mostly stems from a somewhat silly opinion that `types/data` looks ugly and I wouldn't really want us to ever call a Data Type just "Data" because it could easily get very confusing.

Reply:
I think we're going to leave them as is for this RFC, and address properly in the near future when we specify how URIs are going to be formatted. This will possibly be done as part of #408